### ISSUE TITLE
MCD: consider taking action if validation fails multiple times

### ISSUE THREAD

**Issue by sdemos** [openshift/machine-config-operator#99]
a node is validated after being updated but before being marked as completed. this validation normally happens after a reboot. if this validation fails, the update is attempted again, ad infinitum. 

there are three reasons I can think of that would cause validation to fail.
1. a bug in the validation logic.
2. a bug in the update application logic.
3. there is something wrong with the machine that is causing it to either revert the changes or never actually be applied.

for situation 3, it seems like the obvious solution is to set the node status to degraded. after all, that's what that status is essentially for (something's wrong with this machine that we can't fix! help!).

however, it's not possible to reliably distinguish between the situations, and setting the node as degraded for situations 1 and 2 [is not ideal](https://github.com/openshift/machine-config-operator/issues/74).

so I guess the question is, how should we handle this situation? do we mark the node as degraded anyway? do we just keep trying to attempt the update forever? should there be some kind of exponential backoff strategy? do we stop trying after a while and bubble up this issue so a cluster administrator will notice the problem, but without setting the node as degraded (which has other semantic meaning)? thoughts?

**Comment by ashcrow** [comment]
> so I guess the question is, how should we handle this situation? do we mark the node as degraded anyway? do we just keep trying to attempt the update forever?

No! :smiley: 

> should there be some kind of exponential backoff strategy? do we stop trying after a while and bubble up this issue so a cluster administrator will notice the problem, but without setting the node as degraded (which has other semantic meaning)? thoughts?

If we can not validate the host then we are not in the state the cluster wants. My initial thought on the path here would be:

1. Attempt to validate the node
2. It fails. Log. Mark the node as degraded and try again.
3. It validates, log, remove degredation, note the node finished and is once again ready. It fails again. log then try once more
4. It validates, log, remove degredation, note the node finished and is once again ready. It fails again stay in degraded mode, log, and wait for cluster or admin intervention.

I wouldn't be opposed to coming up with a mid-term flag such as `reconciling` while we retry a few times to ensure that it's literally an impossible state for the node to enforce.

**Comment by sdemos** [comment]
I was thinking about the possibility of removing the degraded tag as a solution to this problem, but I don't know what the exact semantics are around tagging a machine with that. My intuition says we shouldn't remove the degraded tag once we've added it, since the use that's always talked about for that tag is for a component to key off of it and reprovision the machine. It seems like if there is a transient issue, we want to try a couple times before marking ourselves degraded if that's how it works (since it would almost definitely happen immediately).

**Comment by ashcrow** [comment]
Fair enough :-)

**Comment by jlebon** [comment]
I like the idea of trying a few times in case of transient flakes before setting degraded and promptly jumping ship (i.e. the daemon has no reason to be running once it's degraded, right?). Though that'd mean we'd need to store the number of times we tried somewhere... another annotation I suppose? Also would need to take care of resetting that counter if the desired config changes while trying again.

There's some overlap here with the automatic rollback discussions happening in FCOS. In a distant future, one can imagine automatic rollback in RHCOS feeding into this. E.g. if GRUB couldn't boot into the new deployment and fell back to the previous one, we should immediately mark the node as degraded.

**Comment by jlebon** [comment]
So, #245 changed this slightly. Now, if the node doesn't validate after rebooting into a new config, we immediately mark degraded instead of retrying. So we've gone from "retry forever" to "retry never". But it's still more correct since we do degrade the node.

Hmm, so I'm wondering if we should close this. I'm not sure if there are "transient issues" here that we should account for when updating the node. E.g. if a service is stomping on our file at startup, then it's just going to keep stomping on our file every reboot, right? It seems like if we successfully wrote files/updated the OS and reboot and the on-disk state doesn't match what we just applied, degrading is the correct thing to do?
